spec/data-model/message.json

```diff
@@ -22,76 +22,92 @@
       },
       "required": ["type", "name"]
     },
-    "value": {
-      "oneOf": [{ "$ref": "#/$defs/literal" }, { "$ref": "#/$defs/variable" }]
+    "option": {
+      "type": "object",
+      "properties": {
+        "name": { "type": "string" },
+        "value": {
+          "oneOf": [{ "$ref": "#/$defs/literal" }, { "$ref": "#/$defs/variable" }]
+        }
+      },
+      "required": ["name", "value"]
     },
 
-    "function": {
+    "function-annotation": {
       "type": "object",
       "properties": {
-        "type": { "const": "function" },
+        "type": { "const": "function-annotation" },
         "kind": { "enum": ["open", "close", "value"] },
         "name": { "type": "string" },
         "options": {
           "type": "array",
-          "items": {
-            "type": "object",
-            "properties": {
-              "name": { "type": "string" },
-              "value": { "$ref": "#/$defs/value" }
-            },
-            "required": ["name", "value"]
-          }
+          "items": { "$ref": "#/$defs/option" }
         }
       },
       "required": ["type", "kind", "name"]
     },
-    "unsupported-expression": {
+    "unsupported-annotation": {
       "type": "object",
       "properties": {
-        "type": { "const": "unsupported-expression" },
+        "type": { "const": "unsupported-annotation" },
         "sigil": {
-          "enum": ["!", "@", "#", "%", "^", "&", "*", "<", ">", "?", "~"]
+          "enum": ["!", "@", "#", "%", "^", "&", "*", "<", ">", "/", "?", "~"]
         },
         "source": { "type": "string" }
       },
       "required": ["type", "sigil", "source"]
     },
-    "annotation": {
-      "oneOf": [
-        { "$ref": "#/$defs/function" },
-        { "$ref": "#/$defs/unsupported-expression" }
-      ]
-    },
 
     "literal-expression": {
       "type": "object",
       "properties": {
+        "type": { "const": "literal-expression" },
```

Comment:
I think adding a `type` at this level would be unnecessary and noisy. It's effectively duplicating information that's already available in the model as
```js
arg ? arg.type : annotation.type
```

Reply:
There are _many_ levels where `type` is unnecessary (e.g., SelectMessage has `selectors` and `variants` while PatternMessage has `pattern`; VariableRef has `name` while Literal has `value`), and in fact the sole exception that I see (InputDeclaration has `value: VariableExpression` rather than `value: VariableRef`) appears to be a bug. So what guidelines should be used? What I have done here is include `type` in the constituents of any union—e.g., `Expression = LiteralExpression | VariableExpression | …`, and all the subtypes can be trivially discriminated by `type`.

> ```js
> arg ? arg.type : annotation.type
> ```

I want to avoid the need for consumers to invent such cleverness, especially since it is not robust in the face of adding private-use expressions.

Reply:
The way I'm looking at this, an expression is a container, i.e. the `{…}` part of the syntax where the `…` might contain an operand, and it might contain an annotation, but in any case the surrounding `{` and `}` stay the same. So the data model should not change or change as little as possible throughout such changes as well.

As it does make sense to encode in the data model the "expression must not be empty" bit, we do need at least two different shapes for the expression contents, i.e. `{ arg: …, func?: … } | { func: … }`. But why go further than that?

I do get your argument about private-use, but if we're renaming its slot in the expression from `func` to `annotation`, then I don't see how it needs to impact the expression-level syntax. OTOH, if we were to continue with a dedicated slot for functions, then I could see something like this making sense:

```ts
type Expression = ValueExpression | FunctionExpression | UnsupportedExpression;

interface ValueExpression {
  type: "expression";
  arg: Literal | VariableRef;
  func?: FunctionRef;
}

interface FunctionExpression {
  type: "expression";
  arg?: null;
  func: FunctionRef;
}

interface UnsupportedExpression {
  type: "unsupported";
  arg?: Literal | VariableRef | null;
  annotation: UnsupportedAnnotation;
}
```
with `UnsupportedAnnotation` as you currently have it.

One previously unstated optimization target I have here is the developer experience of working with MF2 data models, which shows up here in two different ways:
1. If I need to write a script that'll e.g. add a function to all unannotated placeholders matching some variable name, needing to change the expression type in addition to the `func` values feels pretty clumsy.
2. Needing to work with string identifiers that are over 20 characters long is just asking for typos.